### ISSUE TITLE
data can be lost when problems in commitTransaction happen

### ISSUE THREAD

**Issue by rgerhards** [rsyslog/rsyslog#974]
Thanks to Kane Kim for identifying this.

The full begin of this story is on the rsyslog mailing list (e.g. http://lists.adiscon.net/pipermail/rsyslog/2016-March/042379.html), but we should have it on github especially as this may involve coordinated development. Short summary:

During end transaction processing, errors are not handled as good as possible. This is already indicated by a TODO comment: https://github.com/rsyslog/rsyslog/blob/master/action.c#L1236

The comment is partly outdated (see ML), but the core idea is correct. I think it would also be the proper spot to begin to extend things.

The best plan to address this is:

The plan acutally is:
1. get a testbench test to expose that problem -- this greatly eases debugging  line of thought is to update omfile, which is needed anyhow, so that we can induce errors via it
2. add missing functionality to write error file this does not directly address the issue, but is useful and ensures we are on the right path
3. check the end transaction endpoint. After review of old notes and some thinking, I think that      part of the "not yet implemented" comment is actually right. It's not as easy as suggested by it,  but we probably can do better.

During implementation, specifics of engine shutdown need to be considered. Let me stress that I think having a testbench test case is vital in order to apply any changes, as we must make sure this works well on many platforms, which almost rules out manual testing.

I would definitly and strongly recommend to tackle this beast in the order given above. I would also go for low-hanging fruits at first. David Lang has suggested advanced transaction error handling on the ML. This makes sense, but there are many subtle issues to consider. Unfortunately, I cannot elaborate much here, but the bottom line is that we need to have actions explicitely opt-in to use that functionality. So I strongly suggest to do anything advanced only after the base functionality described here is solid.

It should be save if the implementation blocks for the configured number of retries (and watching shutdown state) at the line mentioned above. One might look at the equivalent code path in the non-transactional case.

**Comment by rgerhards** [comment]
moved the error file enhancement to its own issue tracker #1836 so that we can finally close the core issue here.
